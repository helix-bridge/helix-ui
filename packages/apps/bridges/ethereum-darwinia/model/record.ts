import { Network } from 'shared/model';

interface DarwiniaListRes<T> {
  count: number;
  list: T[];
}

export interface Ethereum2DarwiniaRecord {
  address: string;
  amount: string;
  block_num: number;
  block_timestamp: number;
  chain: Network;
  created_at: string;
  currency: string;
  target: string;
  tx: string;
}
export interface Ethereum2DarwiniaRingBurnRecord extends Ethereum2DarwiniaRecord {
  id: number;
}

export type Ethereum2DarwiniaRingBurnHistoryRes<R = Ethereum2DarwiniaRingBurnRecord> = DarwiniaListRes<R>;

export interface Ethereum2DarwiniaRedeemRecord extends Ethereum2DarwiniaRecord {
  darwinia_tx: string;
  deposit: string; // json string
  is_relayed: boolean;
}

export type Ethereum2DarwiniaRedeemHistoryRes<R = Ethereum2DarwiniaRedeemRecord> = DarwiniaListRes<R>;

export interface Darwinia2EthereumRecord {
  account_id: string;
  block_hash: string;
  block_header: string;
  block_num: number;
  block_timestamp: number;
  extrinsic_index: string;
  kton_value: string;
  mmr_index: number;
  mmr_root: string;
  ring_value: string;
  signatures: string;
  target: string;
  tx: string;
}

export type Darwinia2EthereumHistoryRes<R = Darwinia2EthereumRecord> = DarwiniaListRes<R> & {
  implName: string;
  best: number;
  MMRRoot: string;
};

export type Darwinia2EthereumMeta = Pick<Darwinia2EthereumHistoryRes, 'best' | 'MMRRoot'>;
