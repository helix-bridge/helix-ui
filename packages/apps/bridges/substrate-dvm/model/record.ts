interface Transfer {
  amount: string;
  senderId: string;
  timestamp: string;
  recipientId: string;
  section: 'kton' | 'balances';
  method: string;
  // eslint-disable-next-line id-denylist
  block: { blockHash: string; number: number; specVersion: number };
}

export interface Substrate2DVMRecordsRes {
  transfers: {
    totalCount: number;
    nodes: Transfer[];
  };
}

export type DVM2SubstrateRecordsRes = Substrate2DVMRecordsRes;

export type Substrate2DVMRecord = Transfer;

export type DVM2SubstrateRecord = Transfer;
