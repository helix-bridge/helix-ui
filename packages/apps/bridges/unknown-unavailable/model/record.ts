export type Unknown2UnavailableRecord = unknown;
export type Unavailable2UnknownRecord = unknown;
export type Unknown2UnavailableRecordRes = unknown;
export type Unavailable2UnknownRecordRes = unknown;
export type Unknown2UnavailableRecordsRes = unknown;
export type Unavailable2UnknownRecordsRes = unknown;
