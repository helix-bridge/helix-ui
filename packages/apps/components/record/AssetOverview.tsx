import { DATE_FORMAT } from '@helix/shared/config/constant';
import { fromWei, getTimeRange, prettyNumber } from '@helix/shared/utils';
import { Typography } from 'antd';
import { format } from 'date-fns';
import { useTranslation } from 'react-i18next';

export interface AssetOverviewProps {
  amount: string;
  decimals?: number;
  deposit?: { start: number; month: number; deposit_id: string };
  currency: string;
}

export function AssetOverview({ amount, deposit, currency, decimals = 18 }: AssetOverviewProps) {
  const { t } = useTranslation();
  const depositFlag = 'DEPOSIT';

  if (!currency || !amount) {
    return null;
  }

  if (currency.toUpperCase() === depositFlag) {
    const { deposit_id, start, month } = deposit!;
    const { start: startTime, end: endTime } = getTimeRange(start, month);

    return (
      <Typography.Text>
        <span>{deposit_id}</span>
        <p>
          {fromWei({ value: amount }, prettyNumber)} RING ({t('Time')}: {format(startTime, DATE_FORMAT)} -{' '}
          {format(endTime, DATE_FORMAT)})
        </p>
      </Typography.Text>
    );
  }

  return (
    <Typography.Text className="mr-4">
      <span className="mr-2">
        {amount.includes('.')
          ? amount
          : fromWei({ value: amount, decimals }, (value) => prettyNumber(value, { decimal: 9 }))}
      </span>
      <span>{currency}</span>
    </Typography.Text>
  );
}
