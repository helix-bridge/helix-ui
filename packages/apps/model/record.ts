import { Observable } from 'rxjs';
import { RecordRequestParams } from './record-api';

interface SimpleBlock {
  blockHash: string;
  extrinsicHash: string;
  // eslint-disable-next-line id-denylist
  number: number;
  specVersion: number;
}

export interface BridgeDispatchEventRecord {
  data: string; // json string [ChainId, [LaneId, MessageNonce], DispatchResult]
  isSuccess: boolean;
  method:
    | 'MessageRejected'
    | 'MessageVersionSpecMismatch'
    | 'MessageWeightMismatch'
    | 'MessageSignatureMismatch'
    | 'MessageCallDecodeFailed'
    | 'MessageCallRejected'
    | 'MessageDispatchPaymentFailed'
    | 'MessageDispatched';
  block: SimpleBlock;
  index: number;
}

export interface RecordList<T> {
  count: number;
  list: T[];
}

export type FetchRecords<Res extends RecordList<unknown>, Req = RecordRequestParams> = (req: Req) => Observable<Res>;

export interface RecordsHooksResult<T extends RecordList<unknown>, R = RecordRequestParams> {
  fetchIssuingRecords: FetchRecords<T, R>;
  fetchRedeemRecords: FetchRecords<T, R>;
}
