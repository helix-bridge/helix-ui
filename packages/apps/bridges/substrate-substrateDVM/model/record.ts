import { BridgeDispatchEventRecord } from '../../../model';

export interface SubstrateDVM2SubstrateRecord {
  lane_id: string;
  nonce: string;
  request_transaction: string;
  response_transaction: string;
  sender: string;
  recipient: string;
  token: string;
  amount: string;
  // eslint-disable-next-line no-magic-numbers
  result: 0 | 1 | 2;
  start_timestamp: string;
  end_timestamp: string;
}

export interface Substrate2SubstrateDVMRecord {
  laneId: string;
  nonce: string;
  requestTxHash: string;
  responseTxHash: string;
  senderId: string;
  recipient: string;
  token: string;
  amount: string;
  // eslint-disable-next-line no-magic-numbers
  result: 0 | 1 | 2;
  startTimestamp: string;
  endTimestamp: string;
}

export interface SubstrateDVM2SubstrateRecordsRes {
  burnRecordEntities: SubstrateDVM2SubstrateRecord[];
}

export interface SubstrateDVM2SubstrateRecordRes {
  burnRecordEntity: SubstrateDVM2SubstrateRecord;
}

export interface Substrate2SubstrateDVMRecordsRes {
  s2sEvents: {
    totalCount: number;
    nodes: (Substrate2SubstrateDVMRecord & { id: string })[];
  };
}
export interface Substrate2SubstrateDVMRecordRes {
  s2sEvent: Substrate2SubstrateDVMRecord;
}

export interface BridgeDispatchEventRes {
  bridgeDispatchEvent: BridgeDispatchEventRecord;
}
