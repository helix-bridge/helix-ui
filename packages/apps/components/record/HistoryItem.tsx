import { CheckCircleOutlined } from '@ant-design/icons';
import { List, Tag, Tooltip, Typography } from 'antd';
import { formatDistance, fromUnixTime } from 'date-fns';
import Image from 'next/image';
import { PropsWithChildren, ReactNode } from 'react';
import { CrossChainStatus, CrossChainStatusColor } from 'shared/config/constant';
import { Result } from 'shared/model';

interface HistoryItemProps {
  record: {
    result: Result;
    startTime: number;
  };
  token: {
    amount: string;
    symbol: string;
    logo: string;
  };
  process: ReactNode;
}

export const HistoryItem = ({ record, token, children, process }: PropsWithChildren<HistoryItemProps>) => {
  return (
    <List.Item className="flex flex-col lg:flex-row items-start lg:items-center justify-between space-y-2 lg:space-y-0 bg-gray-900 p-2 lg:p-3 mb-2 border border-gray-800">
      <div className="flex space-x-px lg:space-x-2">
        <Tag
          icon={<CheckCircleOutlined />}
          className="flex items-center justify-center pr-0 lg:pr-2 text-white"
          style={{ backgroundColor: CrossChainStatusColor[record.result] }}
        >
          <span className="hidden lg:inline">{CrossChainStatus[record.result]}</span>
        </Tag>

        <div className="flex flex-col">
          <Typography.Text>
            {formatDistance(fromUnixTime(record.startTime), new Date(new Date().toUTCString()), {
              includeSeconds: true,
              addSuffix: true,
            })}
          </Typography.Text>

          {process}
        </div>
      </div>

      <div className="flex items-center gap-4 flex-1 justify-end">
        <Tag
          color="warning"
          className="flex items-center justify-center order-last lg:order-first bg-yellow-500 text-white font-bold text-sm"
          style={{ height: 'fit-content' }}
        >
          Out
        </Tag>

        <div className="flex items-center gap-4 lg:mx-4">
          <Image alt="..." src={`/image/${token.logo}`} width={40} height={40} />

          <div className={` flex flex-col md:${children ? 'w-28' : 'w-12'} truncate`}>
            <Tooltip title={token.amount}>
              <span className="truncate">{token.amount}</span>
            </Tooltip>
            <span className="capitalize">{token.symbol}</span>
          </div>
        </div>

        <div className="mx-4 lg:mx-0 text-pangolin-main">{children}</div>
      </div>
    </List.Item>
  );
};
