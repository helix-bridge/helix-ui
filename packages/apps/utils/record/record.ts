import { isAddress } from 'ethers/lib/utils';
import { RecordStatus } from 'shared/config/constant';
import { CrossChainDirection, HelixHistoryRecord, TokenWithBridgesInfo } from 'shared/model';
import { fromWei, prettyNumber } from 'shared/utils/helper/balance';
import { isPolkadotNetwork } from 'shared/utils/network/network';
import { getOriginChainConfig } from '../network';

export function getTokenConfigFromHelixRecord(
  record: HelixHistoryRecord,
  key: keyof Pick<HelixHistoryRecord, 'feeToken' | 'sendToken' | 'recvToken'>
): TokenWithBridgesInfo | undefined {
  const chain = getOriginChainConfig(record[key === 'recvToken' ? 'toChain' : 'fromChain']);
  const symbol = record[key];

  return chain.tokens.find((item) => {
    if (isAddress(symbol)) {
      return item.address.toLowerCase() === symbol.toLowerCase();
    }
    const isSameSymbol = item.symbol === symbol;

    if (!isSameSymbol) {
      const isSameSymbolCaseInsensitive = item.symbol.toLowerCase() === symbol.toLowerCase();

      if (isSameSymbolCaseInsensitive) {
        console.log(
          `⚠️ Token symbol(${symbol}) from ${record.id} is not consistent with the symbol(${item.symbol}) stored in ${chain.name} configuration!`
        );
        return true;
      }
    }

    return isSameSymbol;
  });
}

export function getReceivedAmountFromHelixRecord(record: HelixHistoryRecord) {
  const { result } = record;
  const fromToken = getTokenConfigFromHelixRecord(record, 'sendToken')!;
  const toToken = getTokenConfigFromHelixRecord(record, 'recvToken');

  let decimals = toToken?.decimals;

  if (result === RecordStatus.refunded) {
    decimals = fromToken.decimals;
  }

  return fromWei({ value: record.recvAmount, decimals }, (val) => prettyNumber(val, { ignoreZeroDecimal: true }));
}

export function getFeeAmountFromHelixRecord(record: HelixHistoryRecord) {
  const { fromChain, feeToken } = record;
  const fromConfig = getOriginChainConfig(fromChain);
  const token = fromConfig.tokens.find((item) => item.symbol === feeToken)!;
  let decimals = token?.decimals;

  if (!decimals) {
    decimals = isPolkadotNetwork(fromChain) && !fromChain.includes('parachain') ? 9 : 18;
  }

  return fromWei({ value: record.fee, decimals });
}

export function getSentAmountFromHelixRecord(record: HelixHistoryRecord) {
  const { fromChain, sendToken, sendAmount } = record;
  const fromToken = getOriginChainConfig(fromChain)!.tokens.find((item) => item.symbol === sendToken);

  return fromWei({ value: sendAmount, decimals: fromToken?.decimals }, (val) =>
    prettyNumber(val, { ignoreZeroDecimal: true })
  );
}

export function getDirectionFromHelixRecord(record: HelixHistoryRecord): CrossChainDirection | null {
  const { fromChain, toChain, sendToken, recvToken, sendAmount, recvAmount } = record;
  const fromConfig = getOriginChainConfig(fromChain);
  const toConfig = getOriginChainConfig(toChain);
  const fromToken = fromConfig.tokens.find((item) => item.symbol.toLowerCase() === sendToken.toLowerCase());
  const toToken = toConfig.tokens.find((item) => item.symbol.toLowerCase() === recvToken.toLowerCase());

  if (fromToken && toToken) {
    return {
      from: { ...fromToken, meta: fromConfig, amount: fromWei({ ...fromToken, amount: sendAmount }) },
      to: { ...toToken, meta: toConfig, amount: fromWei({ ...toToken, amount: recvAmount }) },
    };
  }

  return null;
}

export function isHelixRecord(record: HelixHistoryRecord): boolean {
  return record.bridge.toLowerCase().startsWith('helix');
}

export function isCBridgeRecord(record: HelixHistoryRecord): boolean {
  return record.bridge.toLowerCase().startsWith('cbridge');
}

export function isXCMRecord(record: HelixHistoryRecord): boolean {
  return record.bridge.toLowerCase().startsWith('xcm');
}
