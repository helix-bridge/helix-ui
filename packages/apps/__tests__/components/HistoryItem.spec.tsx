/// <reference types="jest" />

import { HistoryItem } from '../../components/record/HistoryItem';
import { create } from 'react-test-renderer';

describe('<HistoryItem />', () => {
  it('render history item', () => {
    const component = create(
      <HistoryItem
        record={{ result: 1, startTime: 1652689116 }}
        token={{ amount: '8', symbol: 'RING', logo: 'ring.svg' }}
        process={<span>Just for test</span>}
      />
    );

    let tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
