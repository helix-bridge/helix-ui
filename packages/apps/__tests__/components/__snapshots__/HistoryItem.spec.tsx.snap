// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HistoryItem /> render history item 1`] = `
<li
  className="ant-list-item flex flex-col lg:flex-row items-start lg:items-center justify-between space-y-2 lg:space-y-0 bg-gray-900 p-2 lg:p-3 mb-2 border border-gray-800"
>
  <div
    className="flex space-x-px lg:space-x-2"
  >
    <span
      className="ant-tag flex items-center justify-center pr-0 lg:pr-2 text-white"
      style={
        Object {
          "backgroundColor": "#00AA76",
        }
      }
    >
      <span
        aria-label="check-circle"
        className="anticon anticon-check-circle"
        role="img"
      >
        <svg
          aria-hidden="true"
          data-icon="check-circle"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="64 64 896 896"
          width="1em"
        >
          <path
            d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0051.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
          />
          <path
            d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
          />
        </svg>
      </span>
      <span>
        <span
          className="hidden lg:inline"
        >
          success
        </span>
      </span>
    </span>
    <div
      className="flex flex-col"
    >
      <span
        className="ant-typography"
        onClick={null}
        style={
          Object {
            "WebkitLineClamp": undefined,
          }
        }
      >
        about 1 month ago
      </span>
      <span>
        Just for test
      </span>
    </div>
  </div>
  <div
    className="grid grid-cols-2 items-center gap-4 justify-end place-items-center"
  >
    <span
      className="ant-tag ant-tag-warning text-center bg-yellow-500 text-white font-bold w-16"
      style={
        Object {
          "backgroundColor": undefined,
          "height": "fit-content",
        }
      }
    >
      Out
    </span>
    <div
      className="flex items-center gap-4 lg:mx-4"
    >
      <span
        style={
          Object {
            "background": "none",
            "border": 0,
            "boxSizing": "border-box",
            "display": "inline-block",
            "height": "initial",
            "margin": 0,
            "maxWidth": "100%",
            "opacity": 1,
            "overflow": "hidden",
            "padding": 0,
            "position": "relative",
            "width": "initial",
          }
        }
      >
        <span
          style={
            Object {
              "background": "none",
              "border": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": "initial",
              "margin": 0,
              "maxWidth": "100%",
              "opacity": 1,
              "padding": 0,
              "width": "initial",
            }
          }
        >
          <img
            alt=""
            aria-hidden={true}
            src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2740%27%20height=%2740%27/%3e"
            style={
              Object {
                "background": "none",
                "border": 0,
                "display": "block",
                "height": "initial",
                "margin": 0,
                "maxWidth": "100%",
                "opacity": 1,
                "padding": 0,
                "width": "initial",
              }
            }
          />
        </span>
        <img
          alt="..."
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style={
            Object {
              "border": "none",
              "bottom": 0,
              "boxSizing": "border-box",
              "display": "block",
              "height": 0,
              "left": 0,
              "margin": "auto",
              "maxHeight": "100%",
              "maxWidth": "100%",
              "minHeight": "100%",
              "minWidth": "100%",
              "objectFit": undefined,
              "objectPosition": undefined,
              "padding": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": 0,
            }
          }
        />
        <noscript>
          <img
            alt="..."
            data-nimg="intrinsic"
            decoding="async"
            loading="lazy"
            src="/image/ring.svg"
            srcSet="/image/ring.svg 1x, /image/ring.svg 2x"
            style={
              Object {
                "border": "none",
                "bottom": 0,
                "boxSizing": "border-box",
                "display": "block",
                "height": 0,
                "left": 0,
                "margin": "auto",
                "maxHeight": "100%",
                "maxWidth": "100%",
                "minHeight": "100%",
                "minWidth": "100%",
                "objectFit": undefined,
                "objectPosition": undefined,
                "padding": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "width": 0,
              }
            }
          />
        </noscript>
      </span>
      <div>
        <p
          className="truncate md:w-40"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          8
        </p>
        <p
          className="capitalize"
        >
          RING
        </p>
      </div>
    </div>
  </div>
</li>
`;
